lib/elixir/lib/enum.ex

```diff
@@ -1755,8 +1755,12 @@ defmodule Enum do
     :lists.foldl(fun, acc, enumerable)
   end
 
+  def reduce(first..last, acc, fun) when is_function(fun, 2) and first <= last do
+    reduce_range_inc(first, last, acc, fun)
```

Comment:
Can we please match on the ranges just once and do the `first <= last` check inside the function clause? I want to avoid multiple root level checks (even though the compiler likely optimizes it away anyway).